ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Embed/Nary.hs

```diff
@@ -72,8 +61,57 @@ inject' ::
      , Coercible a (f x)
      , Coercible b (f (HardForkBlock xs))
      )
-  => Proxy f -> InjectionContext xs -> Index xs x -> a -> b
-inject' _ startBounds idx = coerce . inject @f startBounds idx . coerce
+  => Proxy f -> InjectionIndex xs x -> a -> b
+inject' _ iidx = coerce . inject @f iidx . coerce
+
+{-------------------------------------------------------------------------------
+  InjectionIndex
+-------------------------------------------------------------------------------}
+
+-- | This data type is isomorphic to an 'Index' that additionally provides a
+-- 'History.Bound' for every era up to and including that index, but none of
+-- the subsequent eras.
+newtype InjectionIndex xs x =
+    InjectionIndex (Telescope (K History.Bound) (State.Current ((:~:) x)) xs)
```

Comment:
A bit tempting to analogously change `Index` to `NS ((:~:) x) xs`, but definitely not in this PR. Maybe as part of #1331. Then, the functions here could be defined using SOP combinators instead of explicit recursion.